File: drum_kit.go
// drum_kit.go — Go версия

package main

import (
	"fmt"
	"math/rand"
	"os"
	"time"
)

type DrumKit struct {
	BPM       int
	BeatLen   float64
	Running   bool
	Pattern   []map[string]int
	Drums     map[string]Drum
}

type Drum struct {
	Name   string
	Key    string
	Symbol string
}

func NewDrumKit(bpm int) *DrumKit {
	return &DrumKit{
		BPM:     bpm,
		BeatLen: 60.0 / float64(bpm),
		Drums: map[string]Drum{
			"bd": {"Бас", "1", "█"},
			"sd": {"Малый", "2", "▓"},
			"hh": {"Хай-хэт", "3", "▒"},
			"t1": {"Том 1", "4", "░"},
			"t2": {"Том 2", "5", "░"},
			"rd": {"Райд", "6", "●"},
			"cr": {"Крэш", "7", "◆"},
		},
	}
}

func (d *DrumKit) generateRockPattern(style string) []map[string]int {
	pattern := []map[string]int{}
	if style == "shuffle" {
		for i := 0; i < 16; i++ {
			beat := map[string]int{"bd": 0, "sd": 0, "hh": 0, "t1": 0, "t2": 0, "rd": 0, "cr": 0}
			if i%8 == 0 || i%8 == 4 {
				beat["bd"] = 1
			}
			if i%8 == 2 || i%8 == 6 {
				beat["sd"] = 1
			}
			if i%2 == 0 {
				beat["hh"] = 1
			}
			pattern = append(pattern, beat)
		}
	} else {
		for i := 0; i < 16; i++ {
			beat := map[string]int{"bd": 0, "sd": 0, "hh": 0, "t1": 0, "t2": 0, "rd": 0, "cr": 0}
			if i%8 == 0 {
				beat["bd"] = 1
				beat["cr"] = 1
			}
			if i%8 == 4 {
				beat["bd"] = 1
			}
			if i%8 == 2 || i%8 == 6 {
				beat["sd"] = 1
			}
			if i%4 == 0 {
				beat["hh"] = 1
			}
			pattern = append(pattern, beat)
		}
	}
	return pattern
}

func (d *DrumKit) playBeat(beat map[string]int) string {
	line := ""
	for _, drum := range d.Drums {
		if beat[drum.Name] == 1 {
			line += "\x1b[32m" + drum.Symbol + "\x1b[0m "
		} else {
			line += "  "
		}
	}
	return line
}

func (d *DrumKit) displayTimeline() {
	for _, drum := range d.Drums {
		fmt.Printf("%s ", drum.Name)
	}
	fmt.Println()
	fmt.Println("─────────────────────────")
}

func (d *DrumKit) run(style string) {
	d.Pattern = d.generateRockPattern(style)
	d.Running = true

	fmt.Printf("\x1b[36m🥁 Rock Drum Kit (Go)\x1b[0m\n")
	fmt.Printf("Темп: %d BPM\n", d.BPM)
	fmt.Printf("Стиль: %s\n", style)
	fmt.Println("Нажмите Ctrl+C для остановки...\n")

	d.displayTimeline()

	beatIndex := 0
	ticker := time.NewTicker(time.Duration(d.BeatLen/2 * float64(time.Second)))
	defer ticker.Stop()

	for range ticker.C {
		if !d.Running {
			break
		}
		beat := d.Pattern[beatIndex]
		line := d.playBeat(beat)
		bar := beatIndex/4 + 1
		beatInBar := beatIndex%4 + 1
		fmt.Printf("%d.%d  %s\n", bar, beatInBar, line)
		beatIndex = (beatIndex + 1) % len(d.Pattern)
	}
}

func main() {
	fmt.Println("🥁 Rock Drum Kit (Go)")
	fmt.Println("1. Автоматический проигрыватель")
	fmt.Print("Ваш выбор (1): ")
	choice := ""
	fmt.Scanln(&choice)

	drum := NewDrumKit(120)

	if choice == "2" {
		// Интерактивный режим (упрощённо)
		fmt.Println("Интерактивный режим не реализован в Go версии")
	} else {
		fmt.Print("Стиль (shuffle/straight): ")
		style := ""
		fmt.Scanln(&style)
		if style != "shuffle" && style != "straight" {
			style = "shuffle"
		}
		drum.run(style)
	}
}

File: drum_kit.rs
// drum_kit.rs — Rust версия

use std::collections::HashMap;
use std::time::{Duration, Instant};
use std::thread;

struct Drum {
    name: String,
    key: String,
    symbol: String,
}

struct DrumKit {
    bpm: u32,
    beat_len: f64,
    running: bool,
    pattern: Vec<HashMap<String, u8>>,
    drums: HashMap<String, Drum>,
}

impl DrumKit {
    fn new(bpm: u32) -> Self {
        let mut drums = HashMap::new();
        drums.insert("bd".to_string(), Drum { name: "Бас".to_string(), key: "1".to_string(), symbol: "█".to_string() });
        drums.insert("sd".to_string(), Drum { name: "Малый".to_string(), key: "2".to_string(), symbol: "▓".to_string() });
        drums.insert("hh".to_string(), Drum { name: "Хай-хэт".to_string(), key: "3".to_string(), symbol: "▒".to_string() });
        drums.insert("t1".to_string(), Drum { name: "Том 1".to_string(), key: "4".to_string(), symbol: "░".to_string() });
        drums.insert("t2".to_string(), Drum { name: "Том 2".to_string(), key: "5".to_string(), symbol: "░".to_string() });
        drums.insert("rd".to_string(), Drum { name: "Райд".to_string(), key: "6".to_string(), symbol: "●".to_string() });
        drums.insert("cr".to_string(), Drum { name: "Крэш".to_string(), key: "7".to_string(), symbol: "◆".to_string() });

        DrumKit {
            bpm,
            beat_len: 60.0 / bpm as f64,
            running: false,
            pattern: Vec::new(),
            drums,
        }
    }

    fn generate_rock_pattern(&mut self, style: &str) {
        let mut pattern = Vec::new();
        if style == "shuffle" {
            for i in 0..16 {
                let mut beat = HashMap::new();
                beat.insert("bd".to_string(), 0);
                beat.insert("sd".to_string(), 0);
                beat.insert("hh".to_string(), 0);
                beat.insert("t1".to_string(), 0);
                beat.insert("t2".to_string(), 0);
                beat.insert("rd".to_string(), 0);
                beat.insert("cr".to_string(), 0);
                if i % 8 == 0 || i % 8 == 4 { beat.insert("bd".to_string(), 1); }
                if i % 8 == 2 || i % 8 == 6 { beat.insert("sd".to_string(), 1); }
                if i % 2 == 0 { beat.insert("hh".to_string(), 1); }
                pattern.push(beat);
            }
        } else {
            for i in 0..16 {
                let mut beat = HashMap::new();
                beat.insert("bd".to_string(), 0);
                beat.insert("sd".to_string(), 0);
                beat.insert("hh".to_string(), 0);
                beat.insert("t1".to_string(), 0);
                beat.insert("t2".to_string(), 0);
                beat.insert("rd".to_string(), 0);
                beat.insert("cr".to_string(), 0);
                if i % 8 == 0 { beat.insert("bd".to_string(), 1); beat.insert("cr".to_string(), 1); }
                if i % 8 == 4 { beat.insert("bd".to_string(), 1); }
                if i % 8 == 2 || i % 8 == 6 { beat.insert("sd".to_string(), 1); }
                if i % 4 == 0 { beat.insert("hh".to_string(), 1); }
                pattern.push(beat);
            }
        }
        self.pattern = pattern;
    }

    fn play_beat(&self, beat: &HashMap<String, u8>) -> String {
        let mut line = String::new();
        for (key, drum) in &self.drums {
            if let Some(&val) = beat.get(key) {
                if val == 1 {
                    line.push_str(&format!("\x1b[32m{}\x1b[0m ", drum.symbol));
                } else {
                    line.push_str("  ");
                }
            } else {
                line.push_str("  ");
            }
        }
        line
    }

    fn display_timeline(&self) {
        for drum in self.drums.values() {
            print!("{} ", drum.name);
        }
        println!("\n─────────────────────────");
    }

    fn run(&mut self, style: &str) {
        self.generate_rock_pattern(style);
        self.running = true;

        println!("\x1b[36m🥁 Rock Drum Kit (Rust)\x1b[0m");
        println!("Темп: {} BPM", self.bpm);
        println!("Стиль: {}", style);
        println!("Нажмите Ctrl+C для остановки...\n");

        self.display_timeline();

        let beat_len_ms = (self.beat_len / 2.0 * 1000.0) as u64;
        let mut beat_index = 0;

        loop {
            let beat = &self.pattern[beat_index];
            let line = self.play_beat(beat);
            let bar = beat_index / 4 + 1;
            let beat_in_bar = beat_index % 4 + 1;
            println!("{}.{}  {}", bar, beat_in_bar, line);
            beat_index = (beat_index + 1) % self.pattern.len();
            thread::sleep(Duration::from_millis(beat_len_ms));
        }
    }
}

fn main() {
    println!("🥁 Rock Drum Kit (Rust)");
    println!("Стиль (shuffle/straight): ");
    let mut style = String::new();
    std::io::stdin().read_line(&mut style).unwrap();
    let style = style.trim();
    let style = if style == "shuffle" || style == "straight" { style } else { "shuffle" };

    let mut kit = DrumKit::new(120);
    kit.run(style);
}
